cmake functions documentation (#591)
STORM-IRIT/Radium-Engine: doc/developer/cmakeutilities.md

When configuring your own target in your `CMakeLists.txt` file, you just need to link with those targets to get access to all the public interface of Radium (include search path, libraries, ...).

The radium package also defines several cmake functions, described below, that you can use to ease the configuration of your application, library or plugin, mainly to install them in a relocatable way while allowing their use from their own build-tree. 

## Function configure_radium_library {#configure_radium_library}
~~~{.cmake}
configure_radium_library( 
    TARGET targetName               # Name of the target to configure as a Radium Library
    FILES file1[file2[file3 ...]]   # list of headers to install
    [TARGET_DIR directoryName]      # Name of the directory where files are installed (default <prefix>/include/<TARGET>)
    [NAMESPACE NameSpace]           # Namespace of the imported target (default Radium)
    [PACKAGE_CONFIG file.cmake.in]  # name of the configure script model for the given target
    [PACKAGE_DIR packageDirName]    # Name of the directory where the cmake config file will be installed (default <prefix>/lib/cmake/Radium)
)
~~~

> This cmake function configure the `<TARGET>` for installation and for further import in client project using `find_package(<TARGET>)`.

Reviewer: ```suggestion
This cmake function configures the `<TARGET>` for installation and for further import in client project using `find_package(<TARGET>)`.
```

In order to use the library either through an imported target from an installed binary or as a project target in another component on the same build tree, this function defines an alias target with the same name than the imported one.

> This function also define the symbol <TARGET>_EXPORTS so that, it could be used to allow symbol import/export from dynamic library.

Reviewer: ```suggestion
This function also defines the symbol <TARGET>_EXPORTS so that, it could be used to allow symbol import/export from dynamic library.
```

This function takes the following parameters :
 
  - `<TARGET>`. The name of the target to configure
  - `<FILES>`. Expected to be public headers, they will be installed in the include directory of the project installation configuration `${CMAKE_INSTALL_PREFIX}/include`.
  - `<TARGET_DIR>`. If given, `<FILES>` will be installed into the `${CMAKE_INSTALL_PREFIX}/include/<TARGET_DIR>` directory. If not, the files will be installed into `${CMAKE_INSTALL_PREFIX}/include/<TARGET>` directory.
  - `<NAMESPACE>`. If given, the imported target will be `<NAMESPACE>::<TARGET>`. If not, the imported target will be `Radium::<TARGET>`
  - `<PACKAGE_CONFIG>`. If given, a configure script, to be used by `find_package`, will be generated. If not, only the exported targets will be generated.
  - `<PACKAGE_DIR>`. If given, the cmake configuration script `<TARGET>Config.cmake` searched by `find_package(<TARGET>)` will be installed in the directory `${CMAKE_INSTALL_PREFIX}/<PACKAGE_DIR>`. If not, the configure script will be installed in the directory `<${CMAKE_INSTALL_PREFIX}/lib/cmake/Radium`.

> In order to allow usage of installed library in all supported systems (Linux, MacOs, Windows), one of the public header, that might be named, e.g. `<TARGET>Macros.hpp` and included all the public or private headers, should contains the following :

Reviewer: ```suggestion
In order to allow usage of installed library in all supported systems (Linux, MacOs, Windows), one of the public header, that might be named, e.g. `<TARGET>Macros.hpp` and included all the public or private headers, should contain the following :
```

~~~{.cpp}
#pragma once
#include <Core/RaCore.hpp>
#if defined <TARGET>_EXPORTS
#    define <TARGET>_API DLL_EXPORT
#else
#    define <TARGET>_API DLL_IMPORT
#endif
~~~

Then, in the library source code, each exported symbol should be prefixed by `<TARGET>_API` e.g.
~~~{.cpp}
#pragma once
#include < <TARGET>/<TARGET>Macros.hpp >
class <TARGET>_API MyAwesomeMaterial : public Ra::Engine::Material
{
    ...
};

<TARGET>_API bool initializeMyLibrary();
~~~

> Assuming a library `MyLib` was configured as the following

Reviewer: ```suggestion
Assuming a library `MyLib` was configured as follows:
```

        PACKAGE_CONFIG ${CMAKE_CURRENT_SOURCE_DIR}/Config.cmake.in
        FILES "${public_headers};${public_inlines}"
)
~~~

This library could be used either from its installed binaries :
~~~{.cmake}
find_package(MyLib) 
...
target_link_libraries(
        MyProject
        PUBLIC
        Radium::Core
        Radium::Engine
        MyLib::MyLib
)
~~~
or from the same buildtree :
~~~{.cmake}
add_subdirectory(MyLibSource) 
...
target_link_libraries(
        MyProject
        PUBLIC
        Radium::Core
        Radium::Engine
        MyLib::MyLib
)
~~~

## Function installResources {#installResources}
~~~{.cmake}
installResources( 
    TARGET targetName               # Name of the target for which resources must be installed                      
    DIRECTORY theResourceDirectory  # location, in the source tree, of the resources 
    [BUILD_LOCATION whereToLink]    # In the build tree, where to link the resources (default ${CMAKE_CURRENT_BINARY_DIR}/../Resources
    [PREFIX TargetResourcePrefix]   # In the install tree, where to install resources (default <empty>).
    [FILES[file1[file2...]]]        # Resources individual files to install
~~~

> When a Radium based component need to access several resources to implement its functionalities, this function install the required resources.

Reviewer: ```suggestion
When a Radium based component needs to access several resources to implement its functionalities, this function installs the required resources.
```

This function takes the following parameters :
 

>   - `<TARGET>`. The name of the target to configure

Reviewer: ```suggestion
  - `<TARGET>`. The name of the target to configure.
```

  - `<DIRECTORY>`. The directory in the source tree that contains the resource to install.
  - `<BUILD_LOCATION>`. If given, `<DIRECTORY>` will be linked in the build tree at this location. If not, the resource directory will be linked to `${CMAKE_CURRENT_BINARY_DIR}/../Resources` directory.

>   - `<PREFIX>`. If given, the resources will be installed into the directory `<install_prefix>/Resources/<PREFIX>/`. If not, the resources will be installed into the directory `<install_prefix>/Resources/` where `<install_prefix>`is the installation directory of the target.

Reviewer: ```suggestion
  - `<PREFIX>`. If given, the resources will be installed into the directory `<install_prefix>/Resources/<PREFIX>/`. If not, the resources will be installed into the directory `<install_prefix>/Resources/` where `<install_prefix>` is the installation directory of the target.
```

  - `<FILES>`. If given, this allow to install only the given files from `<DIRECTORY>`. If not, all the files from `<DIRECTORY>` will be installed.
 

> For a Radium-based software component, resources can be of several type :
>   - Text files, that contains some glsl source code for used shaders.

Reviewer: ```suggestion
For a Radium-based software component, resources can be of several types:
  - Text files, that contain some glsl source code used for shaders.
```

  - Images used internally as textures or source of data.
  - Precomputed data, stored in whatever format.
  

> To access resources, the component might rely on a Ra::Core::Resources::ResourcesLocator object. Such an object allows to access files located in some predefine directories

Reviewer: ```suggestion
To access resources, the component might rely on a Ra::Core::Resources::ResourcesLocator object. Such an object allows to access files located in some predefined directories:
```

  - Radium default resource directory. This directory is located at `<prefix>/Resources` where `<prefix>` is the Radium installation directory.
  - Current executable directory. This directory is where the currently running executable is located.
  

> In order to allow client component to access their own resources, and to allow that either on the build-tree or once the component is installed without having to recompile or modify anything, this function both link (on supported systems, copy elsewhere) and install the component resources at the given places.

Reviewer: ```suggestion
In order to allow client component to access its own resources, and to allow that either on the build-tree or once the component is installed without having to recompile or modify anything, this function both links (on supported systems, copy elsewhere) and installs the component resources at the given places.
```

As the process must be system and component independent it is the responsibility of the client package to configure the proper paths for resources management.

According to the installation layout of the Radium lib, that matches the GNU standard, but also to the MacOs bundle architecture, Resources might be localized relatively to any binary object (library, executable, plugin, ...)

> The Radium installation hierarchy, that embed base libraries, their resources and some executable file is the following.

Reviewer: ```suggestion
The Radium installation hierarchy, that embeds base libraries, their resources and some executable files is the following:
```

```
<prefix>/
├── include/
|   ├── Core/
|   ├── Engine/
|   ...
├── lib/
|   ├── cmake/
|   |   └── Radium/
|   ├──  libCore.so
|   ...
├── bin/
|   ├──  main-app
|   ...
├── Resources/
|   ├── Shaders/
|   ├── Config/
|   ...
```
In this hierarchy, Resources are located in the relative path `../Resources/` from any binary (executable or lib).

> The source code structure of the Radium libraries allow to have the same property on the buildtree.
> When a Radium component need to access some resources, it could either use the Ra::Core::Resources::ResourcesLocator::getRadiumResourcesDir() method or search for the resource relatively to a symbol defined in the component.

Reviewer: ```suggestion
The source code structure of the Radium libraries allows to have the same property on the buildtree.
When a Radium component needs to access some resources, it could either use the Ra::Core::Resources::ResourcesLocator::getRadiumResourcesDir() method or search for the resource relatively to a symbol defined in the component.
```

                    "../" );
// Read files from the found resourcesPath that will prefix the Resource files or directory
~~~

When using the [installResources](#installResources) function to configure the component resources installation, the following call should be made to allow access from both the build tree and the install tree using the above snippet.

~~~{.cmake}
# Assuming MyComponentTarget was configured as a library, plugin or executable, this will configure the resource access in the build tree and the install tree
installTargetResources(
    TARGET MyComponentTarget                            # The name of the target
    DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/MyResource    # Where are the resources in the source tree
    BUILD_LOCATION ${CMAKE_CURRENT_BINARY_DIR}/Resources/MyComponentResources # The resources will be linked here in the buildtree
    PREFIX MyComponentResources # the resources will be installed in <prefix>/Resources/MyComponentResources
)
~~~

# How to write your CMakeLists.txt
When writing your cmake configuration script `CMakeLists.txt`, you might rely on the following guideline to configure the project `ProjectName`.

> Note that this is only guidelines and that you can always write your cmake script from scratch, according you understand what you do.

Reviewer: ```suggestion
Note that these are only guidelines and that you can always write your cmake script from scratch, assuming you understand what you do.
```

cmake_minimum_required(3.13)
# It is recommended to disable in-source build
set(CMAKE_DISABLE_SOURCE_CHANGES ON)
set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)
# if not explicitely set by the user, set the install prefix to a bundle directory in the build tree
if (CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
    set(CMAKE_INSTALL_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/installed-${CMAKE_CXX_COMPILER_ID}" CACHE PATH
        "Install path prefix, prepended onto install directories." FORCE)
    message("Set install prefix to ${CMAKE_INSTALL_PREFIX}")
    set(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT False)
endif ()
# -------------------------------------------------------
# Configure the project
# -------------------------------------------------------
project(ProjectName VERSION 1.0.0)

# The project requires Core and Engine component from Radium
find_package(Radium REQUIRED Core Engine)

...
~~~

Then, according to the type of the targets defined in your project (application, single library, set of libraries, plugin, mix of them), you might configure your project according to the following guidelines.

## Configuring a single installable library

> An installable library could be used to add functionalities built over the Radium libraries and to make these functionalties available to developers.
> Once installed, a library consist in

Reviewer: ```suggestion
An installable library could be used to add functionalities built over the Radium libraries and to make these functionalities available to developers.
Once installed, a library consists in:
```

Once found, the Radium package defines a cmake target for each requested components (`Core`and `Engine` in the example above) in the cmake namespace `Radium`

When configuring your own target in your `CMakeLists.txt` file, you just need to link with those targets to get access to all the public interface of Radium (include search path, libraries, ...).

The radium package also defines several cmake functions, described below, that you can use to ease the configuration of your application, library or plugin, mainly to install them in a relocatable way while allowing their use from their own build-tree. 

The functions defined by the Radium package are the following:
1. Client application configuration
  - [configure_radium_app](#configure_radium_app)
2. Extending Radium through libraries
  - [configure_radium_library](#configure_radium_library).
  - [installTargetResources](#installTargetResources).
  - [configure_radium_package](#configure_radium_package)
  - [radium_exported_resources](#radium_exported_resources) 
3. Extending Radium through plugins 
  - [configure_radium_plugin](#configure_radium_plugin).

## Client application configuration

> ### Function configure_radium_app {#configure_radium_app}

Reviewer 2: ```suggestion
### Function `configure_radium_app` {#configure_radium_app}
```

~~~{.cmake}
configure_radium_app( 
    NAME applicationName                        # The name of the executable target to configure as a relocatable application
    [USE_PLUGINS]                               # Set this if the application uses installed plugins. 
    [RESOURCES ResourceDir1 ResourceDir2 ...]   # List of directories to install as application resources
)
~~~

> This function configure the executable target `<NAME>` for installation so that the installation directory is relocatable and distributable.

Reviewer 2: ```suggestion
This function configures the executable target `<NAME>` for installation so that the installation directory is relocatable and distributable.
```

This function takes the following parameters:

*Parameter name*                            | *Parameter description*
--------------------------------------------|--------------------

> `<NAME>`                                    | The name of the *executable* target to configure and install

Reviewer 2: ```suggestion
`<NAME> applicationName                     | The name of the *executable* target to configure and install
```

`<USE_PLUGINS>`                             | If this option is given, the plugins installed into the Radium bundle at the installation time will be copied into the application bundle.

> `<RESOURCES> ResourceDir1 ResourceDir2 ...` | The optinal list of directories given here will be considered as application resources and installed into the application bundle.

Reviewer 2: ```suggestion
`<RESOURCES> ResourceDir1 ResourceDir2 ...` | Optional list of directories to be considered as application resources and installed into the application bundle.
```

> When installed into a directory `<prefix>`, the application bundle has the following structure on linux and windows or on MacOsX if the executable is not configured as a `MACOSX_BUNDLE`. In this case, the application is a standard `.app` MacOsX application:
> 
> ```
> <prefix>/
> ├── bin/
> |   └──  applicationName
> ├── Resources/
> |   ├── ResourceDir1/
> |   ├── ResourceDir2/
> |   ...
> ```

Reviewer 2: ```suggestion
When installed into a directory `<prefix>`, the application bundle has the following structure on linux, windows and MacOsX (if the executable is not configured as a `MACOSX_BUNDLE`):

~~~
<prefix>/
├── bin/
|   └──  applicationName
├── Resources/
|   ├── ResourceDir1/
|   ├── ResourceDir2/
|   ...
~~~

For `MACOSX_BUNDLE` applications, a standard `.app` MacOsX application is generated.
```

### Limitations
For now (Radium version of August 2020): 
  - MacOsX Bundled application are truly relocatable (on the build system or between compatible systems) if they do not use plugins. 
  - Linux application can be relocated only on the build system.
  - Windows application need further processsing (windeployQt) to be relocatable.

>   

Reviewer 2: ```suggestion
To get relocatable packages, download pre-built binaries here: https://github.com/STORM-IRIT/Radium-Releases/releases
  
```

The functions defined by the Radium package are the following:
1. Client application configuration
  - [`configure_radium_app`](#configure_radium_app).
2. Extending Radium through libraries
  - [`configure_radium_library`](#configure_radium_library).
  - [`installTargetResources`](#installTargetResources).
  - [`configure_radium_package`](#configure_radium_package).
  - [`radium_exported_resources`](#radium_exported_resources).
3. Extending Radium through plugins 
  - [`configure_radium_plugin`](#configure_radium_plugin).
  
## Client application configuration
### Function `configure_radium_app` {#configure_radium_app}
~~~{.cmake}
configure_radium_app( 
    NAME applicationName                        # The name of the executable target to configure as a relocatable application
    [USE_PLUGINS]                               # Set this if the application uses installed plugins. 
    [RESOURCES ResourceDir1 ResourceDir2 ...]   # List of directories to install as application resources
)

> ~~~

Reviewer 2: In the future we could use this http://jesnault.fr/website/document-your-cmake-code-within-doxygen/
(we don't need it for this PR, but it would be ease maintenance).

Author: I was'nt aware doxygen manage cmake scripts. Thanks for the link. As more than this script must be documented, and as this one is documented explicitely in `cmakeutilities.md', I agree that a future PR will be dedicated to doxygen documentation of cmake scripts. Could you post an issue - feature request so that we remember this TODO ?
